odoo/modules/graph.py

```diff
@@ -3,187 +3,142 @@
-        if name in graph:
-            inst = graph[name]
-        else:
-            inst = object.__new__(cls)
-            graph[name] = inst
-        return inst
+    def __len__(self) -> int:
+        return len(self._packages)
 
-    def __init__(self, name, graph, info):
-        self.name = name
-        self.graph = graph
-        self.info = info or getattr(self, 'info', {})
-        if not hasattr(self, 'children'):
-            self.children = []
-        if not hasattr(self, 'depth'):
-            self.depth = 0
-
-    @property
-    def data(self):
-        return self.info
-
-    def add_child(self, name, info):
-        node = Node(name, self.graph, info)
-        node.depth = self.depth + 1
-        if node not in self.children:
-            self.children.append(node)
-        for attr in ('init', 'update', 'demo'):
-            if hasattr(self, attr):
-                setattr(node, attr, True)
-        self.children.sort(key=lambda x: x.name)
-        return node
-
-    def __setattr__(self, name, value):
-        super(Node, self).__setattr__(name, value)
-        if name in ('init', 'update', 'demo'):
-            tools.config[name][self.name] = 1
-            for child in self.children:
-                setattr(child, name, value)
-        if name == 'depth':
-            for child in self.children:
-                setattr(child, name, value + 1)
-
-    def __iter__(self):
-        return itertools.chain(
-            self.children,
-            itertools.chain.from_iterable(self.children)
-        )
-
-    def __str__(self):
-        return self._pprint()
-
-    def _pprint(self, depth=0):
-        s = '%s\n' % self.name
-        for c in self.children:
-            s += '%s`-> %s' % ('   ' * depth, c._pprint(depth+1))
-        return s
-
-    def should_have_demo(self):
-        return (hasattr(self, 'demo') or (self.dbdemo and self.state != 'installed')) and all(p.dbdemo for p in self.parents)
-
-    @property
-    def parents(self):
-        if self.depth == 0:
-            return []
-
-        return (
-            node for node in self.graph.values()
-            if node.depth < self.depth
-            if self in node.children
-        )
+    def __bool__(self) -> bool:
+        return bool(self._packages)
+
+    DEL_MESSAGE = {
+        'depends_loop': (_logger.warning, 'module %s: in a dependency loop, skipped'),
+        'not_installable': (_logger.warning, 'module %s: not installable, skipped'),
+        'missing_depends': (_logger.info, 'module %s: some depends are not loaded, skipped'),
+    }
+
+    def _del_package(self, name: str, del_message: str = '') -> None:
+        if name not in self._packages:
+            return
+        # !! the following warning is very important to make the feature work correctly when there is a 'depends' loop
+        if del_message in self.DEL_MESSAGE:
+            logger, message = self.DEL_MESSAGE[del_message]
+            logger(message, name)
```

Comment:
@rco-odoo if the warning is removed, the `test_graph_order_3` will fail, do you have any idea about it? I guess it is relevant to the python itself. 🤔